### ISSUE TITLE
Is there a reason why the tally() generic doesn't have ... ?

### ISSUE THREAD

**Issue by rpruim** [tidyverse/dplyr#5694]
Making `tally()` generic is breaking {mosaicCore} (as did the introduction of the non-generic `tally()` years ago).  I created a work-around so that {dplyr} and {mosaic} could co-exist.  I need to come up with a new work-around now that `tally()` is generic.  So I've been looking at the new generic.

I'm seeing that the `dplyr::tally()` generic doesn't include `...` as an argument and does include other arguments that could probably be left to the specific methods, contrary to the advice in section 7.1 of https://cran.r-project.org/doc/manuals/R-exts.pdf. I'm wondering why.

**Comment by lionel-** [comment]
If the generic includes arguments, it means that it is intended that all methods implement this behaviour. It would probably make sense to take `...` though, if only to allow for interface extensions.

Note that there is a difference between general purpose generics like `print()` or `summary()` which are unified by a common theme but intended to support arbitrary classes, and specific generics that are part of an API with explicit semantics. That said, the generics in dplyr lie somewhere in between these two cases (since they are defined for a wide range of objects such as data-frame subclasses and lazy databases). Still it does not seem ideal to use the dplyr generic to register a completely unrelated method. 

Oh now I see that it's about supporting `formula` methods. Then it's definitely not the appropriate solution. You should only register a method if you own either the generic or the class. In this case you wouldn't own the generic `dplyr::tally()` nor the `formula` class.

>  I need to come up with a new work-around now that tally() is generic.

I imagine this is about dplyr dispatching unexpectedly on the methods defined in your namespaced when called from there? In that case you could solve it by one of these solutions:

1. Using internal, non-S3 dispatch in your `tally()` function and call `dplyr::tally()` for unknown objects.
2. Renaming your methods so they don't follow the `generic.class` pattern. You can register these methods manually with an `@method generic class` directive (roxygen won't register them automatically with `@export` if they don't follow the pattern).
3. Evaluating in `baseenv()` here https://github.com/ProjectMOSAIC/mosaicCore/blob/1e8cfd29edac4030bd9fb509355663d656655f46/R/tally.R#L151. This should work because you're embedding everything in the call created by `do.call()`, and it will prevent dplyr from finding the methods defined in the calling environment.

I would favour option 1 (unless you're expecting other packages to define methods for your `tally()` generic) because it would simplify the code in your package considerably. Option 2 would require a smaller amount of change to your code. Option 3 would work now but would require careful analysis each time you modify the code.

**Comment by rpruim** [comment]
I came up with another solution, and I'm wondering (a) what you think of it, and (b) which of a couple flavors is best.

1. I renamed my tally generic and its methods `mosaic_tally()`.  That avoids the collision problem.
2. But I want the API to remain the same, so I need to define `tally()`.  Here are some options:

```{r}
tally
function(x, ...) {
  UseMethod("mosaic_tally")
}

tally <- function(x, ...) {
  mosaic_tally(x, ...)
}

tally <- mosaic_tally
```

Is one of these preferred?

`mosaic_tally.tbl`, etc. call `dplyr::tally()` so `mosaicCore` and `dplyr` can co-exist nicely for users.
